[clang] Add test for CWG1807

Endilll opened the pull request:
The test checks that objects in arrays are destructed in reverse order during stack unwinding.

Endilll committed:
[clang] Add test for CWG1807

The test checks that objects in arrays are destructed in reverse order during stack unwinding.

erichkeane commented on clang/test/CXX/drs/dr1807.cpp:
```diff
@@ -0,0 +1,31 @@
+// CHECK:              invoke void @dr1807::S::S(){{.+}}
+// CHECK-NEXT:         {{.+}} unwind label %lpad
+
+// CHECK-LABEL:      lpad:
```
these labels aren't stable/some modes of the build bots remove them, so you'll have to use a wildcard for these.

Endilll replied:
> some modes of the build bots remove them

How unfortunate. How do they look instead? Numbers?

erichkeane replied:
Yeah, they look like numbers.  Also interesting, parameter names disappear completely, and  %0, %1, %2 (up to # params) refer to them withoutnaming them.

nikic replied:
This test is using clang_cc1 with a fixed target. I don't see how different output is possible here.

erichkeane replied:
> This test is using clang_cc1 with a fixed target. I don't see how different output is possible here.

See `-fdiscard-value-names`, which is pretty common on as a default on some downstreams, resource sensitive build bots, and even some folk's local builds.

nikic replied:
The clang driver may or may not pass `-discard-value-names` to cc1 (depending on the build type), but for cc1 the default is always to not discard value names. This (and similar issues) is exactly why cc1 is always used for codegen tests.

Endilll replied:
My takeaway here is that we don't need to worry about names being discarded, but what about stability of names in general? Names I'm hard-coding in this particular test have been stable for 12 years (since 3.0), but I'm not sure if I can generalize this.

erichkeane replied:
> The clang driver may or may not pass `-discard-value-names` to cc1 (depending on the build type), but for cc1 the default is always to not discard value names. This (and similar issues) is exactly why cc1 is always used for codegen tests.

My understanding is it is a build option for the compiler.  I've DEFINITELY gotten build bot complaints for leaving names in in the past.

erichkeane reviewed (commented):
The test should also validate the non-catch case!  The destruction order is also relevant for the 'try' block's cleanup as well.  So there is probably two different 'check' sections here that need to happen.

Endilll commented:
Our C++ DR test suite has little codegen tests (around 3 or 4, depending on how you count). This rather simple DR about order of destruction of stack unwinding aims to establish a precedent for how such test should be written. Specific points I'd like to highlight:
1. Use of `llvm-cxxfilt`. This improves readability, but is rarely used in Clang and LLVM test suites. On this note, I don't really understand why `-n` is important, but it's present in most of the usages of `llvm-cxxfilt`.
2. Hard-coding register names. It doesn't feel good to hard-code them, but again helps readability. I'll leave a rewrite of this test without hard-coded register names. It should be noted that in this particular case names have been stable for 12 years [since 3.0](https://godbolt.org/z/jGW5T5fvx).
3. Richard has left 25 TODOs for codegen tests while working on DRs 0 through 700. I hope those are just TODOs, and not an indication that there are fundamental problems with this kind of tests. CC @zygoloid 

Test without register names hard-coded is going to look along those lines:
```
// CHECK:              invoke void @dr1807::S::S(){{.+}}
// CHECK-NEXT:         {{.+}} unwind label %[[LPAD:[a-z]+]]

// CHECK-LABEL:      [[LPAD]]:
// CHECK:              br {{.+}}, label {{.+}}, label %[[ARRAYDESTROY_BODY:[a-z.]+]]

// CHECK-LABEL:      [[ARRAYDESTROY_BODY]]:
// CHECK:              %[[ARRAYDESTROY_ELEMENT:[a-z.]+]] = getelementptr {{.+}}, i64 -1
// CXX98-NEXT:         invoke void @dr1807::S::~S(){{.*}}%[[ARRAYDESTROY_ELEMENT]]
// SINCE-CXX11-NEXT:   call void @dr1807::S::~S(){{.*}}%[[ARRAYDESTROY_ELEMENT]]
```

Endilll commented:
@erichkeane Defect report reads
> The test should also validate the non-catch case! The destruction order is also relevant for the 'try' block's cleanup as well. So there is probably two different 'check' sections here that need to happen.

So I think while both your points are interesting, they are out of scope of the DR.

erichkeane commented:
> @erichkeane Defect report reads
> 
> > The destruction of fully-constructed array elements when array initialization is terminated by an exception is required by 14.3 [[except.ctor](https://wg21.link/except.ctor#2)] paragraph 2, but the order in which they are to be destroyed is not specified.
> 
> So I think while both your points are interesting, they are out of scope of the DR.

Thanks, Missed that part.  Carry on!

Endilll committed:
Leave a comment in dr18xx.cpp

Endilll committed:
Eliminate the need for readers to guess how the file is named

cor3ntin commented on clang/test/CXX/drs/dr18xx.cpp:
```diff
@@ -164,6 +164,8 @@ void A<double, U>::C<V>::f4() {
 }
 } // namespace dr1804
 
+// dr1807 is in separate file
```
Missing a file?

Endilll replied:
Addressed in the last update.

Endilll committed:
Simplify C++ part of the test

Endilll committed:
Remove `-n` argument passed to llvm-cxxfilt

It doesn't make a difference, and I don't really understand what it does.

Endilll committed:
Replace some hard-coded strings with captures

Label names has to be hard-coded, as variables are not allowed in `CHECK-LABEL` directives.

erichkeane commented on clang/test/CXX/drs/dr1807.cpp:
```diff
@@ -0,0 +1,28 @@
+// CHECK:              br {{.+}}, label {{.+}}, label %arraydestroy.body
+// CHECK-LABEL:      arraydestroy.body:         
+// CHECK:              [[ARRAYDESTROY_ELEMENT:%.*]] = getelementptr {{.+}}, i64 -1
+// CXX98-NEXT:         invoke void @dr1807::S::~S(){{.*}}[[ARRAYDESTROY_ELEMENT]]
```
```suggestion
// CXX98-NEXT:         invoke void @dr1807::S::~S()({{.*}}[[ARRAYDESTROY_ELEMENT]])
```

erichkeane commented on clang/test/CXX/drs/dr1807.cpp:
```diff
@@ -0,0 +1,28 @@
+// CHECK-LABEL:      arraydestroy.body:         
+// CHECK:              [[ARRAYDESTROY_ELEMENT:%.*]] = getelementptr {{.+}}, i64 -1
+// CXX98-NEXT:         invoke void @dr1807::S::~S(){{.*}}[[ARRAYDESTROY_ELEMENT]]
+// SINCE-CXX11-NEXT:   call void @dr1807::S::~S(){{.*}}[[ARRAYDESTROY_ELEMENT]]
```
```suggestion
// SINCE-CXX11-NEXT:   call void @dr1807::S::~S()({{.*}}[[ARRAYDESTROY_ELEMENT]])
```

erichkeane reviewed (approved):
1 nit for readability sake, else is good enough.

Endilll committed:
Add parens to `invoke` and `call` matchers

Endilll commented:
Following the suggestion from @nikic, I prepared two variants of the same test, one hard-written, and one using `update_cc_test_checks.py` (with heavy manual editing afterwards). Hand-written:
```
// CHECK-LABEL:      define dso_local void @dr1807::f()
// CHECK:              invoke void @dr1807::S::S(){{.+}}
// CHECK-NEXT:         {{.+}} unwind label %lpad
// CHECK-LABEL:      lpad:
// CHECK:              br {{.+}}, label {{.+}}, label %arraydestroy.body
// CHECK-LABEL:      arraydestroy.body:
// CHECK:              [[ARRAYDESTROY_ELEMENT:%.*]] = getelementptr {{.+}}, i64 -1
// CXX98-NEXT:         invoke void @dr1807::S::~S(){{.*}}[[ARRAYDESTROY_ELEMENT]]
// SINCE-CXX11-NEXT:   call void @dr1807::S::~S(){{.*}}[[ARRAYDESTROY_ELEMENT]]
```
Script:
```
// CHECK-LABEL:      define dso_local void @dr1807::f()(
// CHECK:              invoke void @dr1807::S::S()(ptr noundef nonnull align 1 dereferenceable(1) [[ARRAYCTOR_CUR:%.*]])
// CHECK-NEXT:                 to label [[INVOKE_CONT:%.*]] unwind label [[LPAD:%.*]]
// CHECK-LABEL:      lpad:
// CHECK:              br i1 [[ARRAYDESTROY_ISEMPTY:%.*]], label [[ARRAYDESTROY_DONE2:%.*]], label [[ARRAYDESTROY_BODY:%.*]]
// CHECK-LABEL:      arraydestroy.body:
// CHECK:              [[ARRAYDESTROY_ELEMENT]] = getelementptr inbounds %"struct.dr1807::S", ptr [[ARRAYDESTROY_ELEMENTPAST]], i64 -1
// CXX98-NEXT:         invoke void @dr1807::S::~S()(ptr noundef nonnull align 1 dereferenceable(1) [[ARRAYDESTROY_ELEMENT]])
// CXX98-NEXT:                 to label [[INVOKE_CONT1]] unwind label [[TERMINATE_LPAD:%.*]]
// SINCE-CXX11-NEXT:   call void @dr1807::S::~S()(ptr noundef nonnull align 1 dereferenceable(1) [[ARRAYDESTROY_ELEMENT5:.*]]) #[[ATTR2:.*]]
```
In my opinion, script captures way too much, diluting the point of the test. For example, `i64 -1` is a crucial part of the test, but it's much easier to miss it in lengthy script output. (Searching for it might be a good exercise for the reader.) I'd also like to note that target audience of DR tests is likely to be far less experienced in reading LLVM IR than regular users of the script, so it's much less obvious for them to know which parts of the output to pay attention to. On the other hand wildcards help hide uninteresting parts of the codegen output, highlighting what's important.

This brings me to conclusion that hand-written style of codegen tests presented above is a better fit for purposes of C++ defect report codegen tests.

CC @AaronBallman

AaronBallman commented:
> This brings me to conclusion that hand-written style of codegen tests presented above is a better fit for purposes of C++ defect report codegen tests.

I share that conclusion. I think the hand-written IR is easier for people to reason because it strips out unrelated details that make the test somewhat more brittle than it needs to be for DR testing. I agree that we should use the script whenever possible, but DR testing is a bit different from typical codegen testing IMO. So I'm fine with the hand-crafted IR unless I'm missing something.

Endilll committed:
Merge remote-tracking branch 'upstream/main' into cwg1807

erichkeane commented:
> > This brings me to conclusion that hand-written style of codegen tests presented above is a better fit for purposes of C++ defect report codegen tests.
> 
> I share that conclusion. I think the hand-written IR is easier for people to reason because it strips out unrelated details that make the test somewhat more brittle than it needs to be for DR testing. I agree that we should use the script whenever possible, but DR testing is a bit different from typical codegen testing IMO. So I'm fine with the hand-crafted IR unless I'm missing something.

Agreed, I find that using the script makes the tests WAY too fragile and hides the important parts of the test.  I DISAGREE we should use the script when possible, as I basically never see a time where it improves a CFE test.

nico commented:
This breaks tests on macOS: http://45.33.8.238/macm1/76983/step_7.txt

Please take a look and revert for now if it takes a while to fix.

Endilll commented:
@nico I tried to address this in 30da0f5a359ab4a684c5fdf0f4dbed20bae10f99

mysterymath commented:
It looks like 30da0f5  didn't fully fix this; Fuchsia's Mac toolchain builder just completed a run with this change that still fails that test:
https://luci-milo.appspot.com/ui/p/fuchsia/builders/toolchain.ci/clang-mac-x64/b8758543058445552977/test-results?sortby=&groupby=
https://luci-milo.appspot.com/ui/p/fuchsia/builders/toolchain.ci/clang-mac-x64/b8758543058445552977/blamelist

Endilll commented:
@mysterymath It's a bit of a shame to admit that I missed the first half of RUN lines in the previous follow-up. Should be fixed now by a7588bb9bab43420f1c2642c80489f74af88f855
